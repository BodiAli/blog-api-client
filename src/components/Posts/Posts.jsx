import { useEffect, useReducer, useState } from "react";
import { toast } from "react-toastify";
import Card from "../Card/Card";
import Loader from "../Loader/Loader";
import styles from "./Posts.module.css";

export default function Posts() {
  const [posts, dispatch] = useReducer(reducer, []);
  const [totalePages, setTotalPages] = useState(0);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    async function fetchPosts() {
      try {
        setLoading(true);
        const res = await fetch(`${import.meta.env.VITE_SERVER_URL}/posts`);

        if (!res.ok) {
          throw new Error("Failed to fetch posts");
        }

        const { posts, pages } = await res.json();

        dispatch({ type: "set-posts", posts });
        setTotalPages(pages);
      } catch (error) {
        toast.error(error.message);
      } finally {
        setLoading(false);
      }
    }
    fetchPosts();
  }, []);

  if (loading) return <Loader />;

  return (
    <main>
      {/* <Search/> */}
      {posts.length === 0 ? (
        <p className={styles.noPosts}>No published posts yet!</p>
      ) : (
        <>
          <div className={styles.postsContainer}>
            {posts.map((post) => {
              return <Card key={post.id} post={post} />;
            })}
          </div>
          {/* <div className={styles.pagination}>
            <button
              className={currentPage <= 1 ? styles.disabled : ""}
              disabled={currentPage <= 1}
              onClick={() => {
                navigate(`?page=${currentPage - 1}`, { viewTransition: true });
              }}
            >
              Back
            </button>
            {Array.from({ length: totalPages }, (_val, i) => i + 1).map((pageNumber) => {
              return (
                <button
                  onClick={() => {
                    navigate(`?page=${pageNumber}`, { viewTransition: true });
                  }}
                  key={pageNumber}
                  className={currentPage === pageNumber ? styles.active : ""}
                >
                  {pageNumber}
                </button>
              );
            })}
            <button
              className={currentPage >= totalPages ? styles.disabled : ""}
              disabled={currentPage >= totalPages}
              onClick={() => {
                navigate(`?page=${currentPage + 1}`, { viewTransition: true });
              }}
            >
              Next
            </button>
          </div> */}
        </>
      )}
    </main>
  );
}

function reducer(state, action) {
  switch (action.type) {
    case "set-posts":
      return action.posts;
    default:
      throw new Error("Invalid type");
  }
}
