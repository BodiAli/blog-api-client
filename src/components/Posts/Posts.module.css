.main {
  padding: 1rem;
  font-family: "roboto", Arial, sans-serif;
}

.noPosts {
  text-align: center;
  font-size: 1.4rem;
  margin-top: 1rem;
}

.postsContainer {
  display: flex;
  flex-direction: column;
  gap: 2rem;
}

.postsContainer hr {
  width: 100%;
}

.pagination {
  margin-top: 2rem;
  display: flex;
  justify-content: center;
  gap: 1.5rem;
  flex-wrap: wrap;
}

.pagination > button {
  border: none;
  border-radius: 5px;
  color: white;
  padding: 0.4rem 0.8rem;
  font-size: 1rem;
  cursor: pointer;
  transition: background-color 200ms;
  background-color: rgb(72, 130, 180);
  &:hover:not(.disabled) {
    background-color: rgb(102, 148, 188);
  }
}

.pagination > .active {
  border: 2px solid black;
}

.pagination > .disabled {
  background-color: rgba(72, 130, 180, 0.463);
  cursor: not-allowed;
}
