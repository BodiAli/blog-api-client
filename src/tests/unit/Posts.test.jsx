import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createRoutesStub } from "react-router";
import Posts from "../../components/Posts/Posts";

vi.stubGlobal("fetch", vi.fn());

const fakeResponse = {
  posts: [
    {
      id: "57c20dfb-8135-42eb-8abe-0bffbaffd668",
      userId: 1,
      title: "16",
      content: "<p>Content</p>",
      published: true,
      likes: 0,
      createdAt: "2025-05-01T10:44:36.648Z",
      updatedAt: "2025-05-01T10:44:36.648Z",
      imgUrl: null,
      User: {
        id: 1,
        firstName: "Bodi",
        lastName: "Ali",
        profileId: 1,
        Profile: {
          profileImgUrl: "img-url",
        },
      },
      Topics: [
        {
          id: 1,
          name: "Topic 1",
        },
      ],
    },
  ],
  pages: 1,
};

const Stub = createRoutesStub([
  {
    path: "/",
    Component: Posts,
  },
]);

beforeEach(() => {
  fetch.mockImplementation(() => {
    return Promise.resolve({
      ok: true,
      json: () => Promise.resolve(fakeResponse),
    });
  });
  fetch.mockClear();
});

describe("Posts component", () => {
  test("Should render Search component, all posts with Card component, and pagination", async () => {
    let container;

    await waitFor(() => {
      const result = render(<Stub />);
      container = result.container;
    });

    expect(container).toMatchSnapshot();
  });

  test("typing in search input triggers fetch with correct topic query", async () => {
    await waitFor(() => {
      render(<Stub />);
    });

    expect(fetch).toHaveBeenCalledOnce();

    const searchInput = screen.getByRole("searchbox", { name: /search/i });
    expect(searchInput).toHaveValue("");

    await userEvent.type(searchInput, "Topic 1");

    const lastFetchCall = fetch.mock.lastCall[0];

    expect(lastFetchCall).toMatch(/topic=Topic 1/i);
    expect(searchInput).toHaveValue("Topic 1");
  });

  test("Clicking pagination buttons will call fetch with correct page number", async () => {
    const user = userEvent.setup();

    fetch.mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        status: 200,
        json: () =>
          Promise.resolve({
            ...fakeResponse,
            pages: 2,
          }),
      });
    });

    await waitFor(() => {
      render(<Stub />);
    });

    /* Query buttons after click to prevent clicking a stale DOM node because after react-router
   navigates to a new page query, the Posts component will re-render which unmounts the old buttons and inserts
   a brand new one */
    const secondPageButton = screen.getByRole("button", { name: "2" });
    await user.click(secondPageButton);

    expect(fetch.mock.lastCall[0]).toMatch(/page=2/);

    const firstPageButton = screen.getByRole("button", { name: "1" });
    await user.click(firstPageButton);

    expect(fetch.mock.lastCall[0]).toMatch(/page=1/);

    const nextPageButton = screen.getByRole("button", { name: "Next" });
    await user.click(nextPageButton);

    expect(fetch.mock.lastCall[0]).toMatch(/page=2/);

    const backPageButton = screen.getByRole("button", { name: "Back" });
    await user.click(backPageButton);

    expect(fetch.mock.lastCall[0]).toMatch(/page=1/);
  });

  test("if currentPage is on 1 the back button will be disabled", async () => {
    fetch.mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        status: 200,
        json: () =>
          Promise.resolve({
            ...fakeResponse,
            pages: 2,
          }),
      });
    });

    await waitFor(() => {
      render(<Stub />);
    });

    expect(screen.getByRole("button", { name: "Back" })).toBeDisabled();
  });

  test("if currentPage is equal to or bigger than the last page then the next button will be disabled", async () => {
    const user = userEvent.setup();

    fetch.mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        status: 200,
        json: () =>
          Promise.resolve({
            ...fakeResponse,
            pages: 2,
          }),
      });
    });

    await waitFor(() => {
      render(<Stub />);
    });

    const secondPageButton = screen.getByRole("button", { name: "2" });

    await user.click(secondPageButton);

    expect(screen.getByRole("button", { name: "Next" })).toBeDisabled();
  });
});
