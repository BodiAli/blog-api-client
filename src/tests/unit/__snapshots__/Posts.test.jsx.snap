// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Posts component > Should render Search component, all posts with Card component, and pagination 1`] = `
<div>
  <main
    class="_main_a4ed0b"
  >
    <div
      class="_container_a9d8a5"
    >
      <label
        for="search"
      >
        Search by topic
      </label>
      <input
        id="search"
        type="search"
        value=""
      />
    </div>
    <div
      class="_postsContainer_a4ed0b"
    >
      <div
        class="_card_1f2933"
      >
        <div
          class="_text_1f2933"
        >
          <div
            class="_textContainer1_1f2933"
          >
            <img
              alt="Bodi Ali's profile picture"
              class="_profilePic_1f2933"
              src="img-url"
            />
            <p
              class="_name_1f2933"
            >
              Bodi
               
              Ali
            </p>
            <p
              class="_createdAt_1f2933"
            >
              May 1, 2025
            </p>
          </div>
          <div
            class="_textContainer2_1f2933"
          >
            <p
              class="_title_1f2933"
            >
              16
            </p>
          </div>
          <div
            class="_textContainer3_1f2933"
          >
            <p>
              Content
            </p>
          </div>
          <div
            class="_textContainer4_1f2933"
          >
            <p>
              Topic 1
            </p>
          </div>
          <div
            class="_textContainer5_1f2933"
          >
            <p>
              <strong>
                Likes: 
              </strong>
              0
            </p>
          </div>
        </div>
        <div
          class="_imageContainer_1f2933"
        >
          <img
            alt="Post cover"
            src="/src/assets/images/no-image.svg"
          />
        </div>
      </div>
      <hr />
    </div>
    <div
      class="_pagination_a4ed0b"
    >
      <button
        class="_disabled_a4ed0b"
        disabled=""
      >
        Back
      </button>
      <button
        class="_active_a4ed0b"
      >
        1
      </button>
      <button
        class="_disabled_a4ed0b"
        disabled=""
      >
        Next
      </button>
    </div>
  </main>
</div>
`;
